File: tests/tests/telecom-apps/Utils/src/android/telecom/cts/apps/NotificationUtils.java
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package android.telecom.cts.apps;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Person;
import android.content.Context;
import android.content.Intent;
import android.service.notification.StatusBarNotification;
import android.util.Log;

public class NotificationUtils {
    private static final String TAG = NotificationUtils.class.getSimpleName();
    private static final String CALL_STYLE_INITIAL_TEXT = "New Call";
    private static final String CALL_STYLE_ONGOING_TEXT = "Ongoing Call";
    private static final String CALL_STYLE_TITLE = "Telecom CTS Test App";

    public static Notification createCallStyleNotification(
            Context context,
            String channelId,
            String callerName,
            boolean isOutgoing) {
        PendingIntent fullScreenIntent = createPendingIntent(context);
        Person person = new Person.Builder().setName(callerName).setImportant(true).build();
        return new Notification.Builder(context, channelId)
                .setContentText(CALL_STYLE_INITIAL_TEXT)
                .setContentTitle(CALL_STYLE_TITLE)
                .setSmallIcon(android.R.drawable.sym_def_app_icon)
                .setStyle(getCallStyle(isOutgoing, person, fullScreenIntent))
                .setFullScreenIntent(fullScreenIntent, true)
                .setOngoing(isOutgoing)
                .build();
    }

    private static Notification.CallStyle getCallStyle(
            boolean isOutgoing,
            Person person,
            PendingIntent fullScreenIntent) {
        if (isOutgoing) {
            return Notification.CallStyle.forOngoingCall(
                    person,
                    fullScreenIntent);
        } else {
            return Notification.CallStyle.forIncomingCall(
                    person,
                    fullScreenIntent,
                    fullScreenIntent);
        }
    }

    public static void updateNotificationToOngoing(Context context,
            String channelId,
            String callerName,
            int notificationId) {
        PendingIntent ongoingCall = createPendingIntent(context);

        Notification callStyleNotification = new Notification.Builder(context,
                channelId)
                .setContentText(CALL_STYLE_ONGOING_TEXT)
                .setContentTitle(CALL_STYLE_TITLE)
                .setSmallIcon(android.R.drawable.sym_def_app_icon)
                .setStyle(Notification.CallStyle.forOngoingCall(
                        new Person.Builder().setName(callerName).setImportant(true).build(),
                        ongoingCall)
                )
                .setFullScreenIntent(ongoingCall, true)
                .setOngoing(true)
                .build();

        NotificationManager notificationManager =
                context.getSystemService(NotificationManager.class);

        notificationManager.notify(notificationId, callStyleNotification);
    }

    private static PendingIntent createPendingIntent(Context context) {
        return PendingIntent.getActivity(context, 0,
                new Intent(context, context.getClass()), PendingIntent.FLAG_IMMUTABLE);
    }

    public static void clearNotification(Context c, int notificationId) {
        NotificationManager notificationManager = c.getSystemService(NotificationManager.class);
        if (notificationManager != null) {
            notificationManager.cancel(notificationId);
        }
    }

    public static void deleteNotificationChannel(Context c, String notificationChannelId){
        NotificationManager notificationManager = c.getSystemService(NotificationManager.class);
        try {
            // deleteNotificationChannel has sometimes thrown exceptions causing test failures
            notificationManager.deleteNotificationChannel(notificationChannelId);
        }
        catch (Exception e){
            Log.i(TAG, String.format("onUnbind: hit exception=[%s] while deleting the"
                            + " call channel with id=[%s]", e, notificationChannelId));
        }
    }

    public static boolean isTargetNotificationPosted(Context c, int targetNotificationId) {
        NotificationManager notificationManager = c.getSystemService(NotificationManager.class);
        StatusBarNotification[] sbinArray = notificationManager.getActiveNotifications();
        for (StatusBarNotification sbn : sbinArray) {
            if (sbn.getId() == targetNotificationId) {
                return true;
            }
        }
        return false;
    }
}

File: tests/tests/permission/src/android/permission/cts/NotificationUtils.kt
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.permission.cts

import android.service.notification.StatusBarNotification
import org.junit.Assert
import java.util.concurrent.TimeUnit

/**
 * Utility methods to interact with NotificationManager through the CTS NotificationListenerService
 * to get or clear notifications.
 */
object NotificationUtils {

    private val NOTIFICATION_CANCELLATION_TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(5)

    /**
     * Get a notification listener notification that is currently visible.
     *
     * @param cancelNotification if `true` the notification is canceled inside this method
     * @return The notification or `null` if there is none
     */
    @JvmStatic
    @Throws(Throwable::class)
    fun getNotificationForPackageAndId(
        pkg: String,
        id: Int,
        cancelNotification: Boolean
    ): StatusBarNotification? {
        val notificationService = NotificationListener.getInstance()
        val notifications: List<StatusBarNotification> = getNotificationsForPackage(pkg)
        if (notifications.isEmpty()) {
            return null
        }
        for (notification in notifications) {
            if (notification.id == id) {
                if (cancelNotification) {
                    clearNotification(notification)
                }
                return notification
            }
        }
        return null
    }

    /**
     * Clears all currently visible notifications for a specified package.
     */
    @JvmStatic
    @Throws(Throwable::class)
    fun clearNotificationsForPackage(pkg: String) {
        val notifications: List<StatusBarNotification> = getNotificationsForPackage(pkg)
        if (notifications.isEmpty()) {
            return
        }

        clearNotifications(notifications)
    }

    /** Clears the specified notification and ensures (asserts) it was removed */
    @JvmStatic
    @Throws(Throwable::class)
    fun clearNotification(notification: StatusBarNotification) {
        val notificationService = NotificationListener.getInstance()
        notificationService.cancelNotification(notification.key)

        // Wait for notification to get canceled
        TestUtils.eventually({
            Assert.assertFalse(
                listOf(*notificationService.activeNotifications)
                    .contains(notification)
            )
        }, NOTIFICATION_CANCELLATION_TIMEOUT_MILLIS)
    }

    private fun clearNotifications(notifications: List<StatusBarNotification>) {
        val notificationService = NotificationListener.getInstance()
        notifications.forEach { notificationService.cancelNotification(it.key) }

        // Wait for notification to get canceled
        TestUtils.eventually({
            val activeNotifications: List<StatusBarNotification> =
                listOf(*notificationService.activeNotifications)
            Assert.assertFalse(
                activeNotifications.any { notifications.contains(it) }
            )
        }, NOTIFICATION_CANCELLATION_TIMEOUT_MILLIS)
    }

    /**
     * Get all notifications associated with a given package that are currently visible.
     * @param pkg Package for which to filter the notifications by
     * @return [List] of [StatusBarNotification]
     */
    @Throws(Exception::class)
    private fun getNotificationsForPackage(pkg: String): List<StatusBarNotification> {
        val notificationService = NotificationListener.getInstance()
        val notifications: MutableList<StatusBarNotification> = ArrayList()
        for (notification in notificationService.activeNotifications) {
            if (notification.packageName == pkg) {
                notifications.add(notification)
            }
        }
        return notifications
    }
}